dispatch/filter.go

```diff
@@ -0,0 +1,243 @@
+/**
+ * Copyright 2020 Comcast Cable Communications Management, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package dispatch
+
+import (
+	"bytes"
+	"context"
+	"crypto/hmac"
+	"crypto/sha1"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/go-kit/kit/metrics"
+	"github.com/xmidt-org/mimisbrunnr/model"
+	"github.com/xmidt-org/webpa-common/logging"
+	"github.com/xmidt-org/webpa-common/semaphore"
+	"github.com/xmidt-org/wrp-go/v2"
+)
+
+type F interface {
+	Start(context.Context) error
+	Filter(event *wrp.Message)
+	Stop(context.Context) error
+}
+
+type Filter struct {
+	DeviceId     string
+	Norn         model.Norn
+	FilterQueue  atomic.Value
+	Wg           sync.WaitGroup
+	Measures     *Measures
+	Mutex        sync.RWMutex
+	DropUntil    time.Time
+	MaxWorkers   int
+	Workers      semaphore.Interface
+	FailureMsg   FailureMessage
+	CutOffPeriod time.Duration
+	Logger       log.Logger
+	Sender       func(*http.Request) (*http.Response, error)
+	QueueSize    int
+	Dispatcher   D
+}
+
+type FilterConfig struct {
+	QueueSize    int
+	SenderConfig SenderConfig
+}
+
+type FilterDispatcher struct {
+	Norn       model.Norn
+	Dispatcher D
+	Filter     F
+}
+
+func NewFilter(fc FilterConfig, deviceID string, dispatcher D, norn model.Norn, transport http.RoundTripper) (F, error) {
+	if deviceID == "" {
+		return nil, fmt.Errorf("invalid deviceID")
+	}
+	filter := Filter{
+		DeviceId:   deviceID,
+		Dispatcher: dispatcher,
+		FailureMsg: FailureMessage{
+			Text:         FailureText,
+			CutOffPeriod: fc.SenderConfig.CutOffPeriod.String(),
+		},
+		Norn: norn,
+		Sender: (&http.Client{
+			Transport: transport,
+		}).Do,
+	}
+	filter.FilterQueue.Store(make(chan *wrp.Message, fc.QueueSize))
+	return &filter, nil
+}
+
+func (f Filter) Start(_ context.Context) error {
+	f.Wg.Add(1)
+	go f.sendEvents()
+	return nil
+
+}
+
+func (f Filter) Stop(_ context.Context) error {
+	close(f.FilterQueue.Load().(chan *wrp.Message))
+	f.Mutex.Lock()
+	f.Measures.EventQueueDepthGauge.Set(0.0)
+	f.Mutex.Unlock()
+	f.Wg.Wait()
+	return nil
+}
+
+func (f *Filter) Filter(event *wrp.Message) {
+	if f.DeviceId == f.Norn.DeviceID {
```

Comment:
These are the same thing....we should compare the event's device ID to the norn's deviceID.